Step2 : 문자열 계산기

jysohn0825 opened the pull request:
Q1. 테스트 코드를 위해 private 메소드를 생성하지 않았는데, 테스트 코드와 접근 제어자를 어떤 관점으로 바라봐야 할까요?
Q2. 테스트 코드를 처음 작성해 보는데, 단위 테스트를 작성해야 하는 범위가 어디까지 일까요?
Q3. 패키지 및 파일을 분리하여 작성(e.g. Calculator, Number, Operator)하는 기준이 따로 있을까요?

jysohn0825 committed:
kotlin basic grammar

jysohn0825 committed:
Add Calculator Production Code

without test code

jysohn0825 committed:
Add Test Code

add test code
optimizing production & test code

jysohn0825 committed:
Merge branch 'jysohn0825' into step2

kyucumber commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+package calculator
```
> Q1. 테스트 코드를 위해 private 메소드를 생성하지 않았는데, 테스트 코드와 접근 제어자를 어떤 관점으로 바라봐야 할까요?

private 메소드를 호출하는 메소드 테스트하는 형태로 사용하시고 테스트를 위해 접근 제어자를 푸는 것은 지양하시는게 좋습니다. 만약에 퍼블릭 메소드로 private 로직을 모두 테스트하기 어려울 정도로 복잡하거나 한 로직이라면 분리가 필요하진 않은지 한번 더 고민해보시면 좋겠네요.

kyucumber commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+        var result = expressions[0].toLong()
+        for (index in 1 until expressions.size step 2) {
+            result = calculate(
+                operator = Operator.find(expressions[index]),
```
operator는 kotlin의 예약어입니다. 가급적 다른 단어를 이용해 변수명을 지어주시면 좋을 것 같네요.

kyucumber commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+package calculator
+
```
> Q2. 테스트 코드를 처음 작성해 보는데, 단위 테스트를 작성해야 하는 범위가 어디까지 일까요?

작성하신 모든 클래스 및 메소드가 단위 테스트의 대상이 됩니다. 이외의 외부 라이브러리의 메소드 등은 해당 라이브러리에서 충분히 검증이 되었다고 가정하고 테스트하실 필요는 없어요.

kyucumber replied:
> Q3. 패키지 및 파일을 분리하여 작성(e.g. Calculator, Number, Operator)하는 기준이 따로 있을까요?

명확한 기준은 없지만 클래스 하나는 하나의 파일에 정의해주시고 클래스를 작성할때는 단일 책임 원칙을 지키고 결합도를 낮추고 응집도를 높이는 관점에서 작성해주시면 됩니다.

하나의 클래스가 여러 책임을 동시에 수행하고 다른 클래스의 변경에 따라 많은 영향을 받는 경우면 분리가 필요하지 않은지 고민해보세요.

kyucumber commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+
+class Calculator {
+
+    fun calculator(expression: String): Long {
```
Calculator는 명사의 형태라 메소드에 적합한 네이밍이 아닌걸로 보여요. 함수는 동사구로 네이밍해주세요~

kyucumber commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+
+    private fun String.splitByBlank() = this.split(" ")
+
+    private fun validCheck(expression: String?) {
+        if (expression.isNullOrBlank()) throw IllegalArgumentException("Empty Expression")
+        if (isOperatorValid(expression)) throw IllegalArgumentException("Invalid Operator")
+    }
```
코틀린의 require을 활용해보세요.

관련 내용은 [이펙티브 코틀린 아이템 05. 예외를 활용해 코드에 제한을 걸어라](https://kyucumber.github.io/posts/book/effective-kotlin/effective-kotlin-item-05)를 참고해보세요.

kyucumber commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,14 @@
+package calculator
+
+enum class Operator(val operator: String) {
+    PLUS("+"), MINUS("-"), DIVISION("/"), MULTIPLE("*");
```
Operator enum이 각각을 연산하기 위한 함수를 가지면 아래의 when 구문을 없앨 수 있을까요?

```kotlin
when (operator) {
        Operator.PLUS -> number1.plus(number2)
        Operator.MINUS -> number1.minus(number2)
        Operator.DIVISION -> number1.div(number2)
        Operator.MULTIPLE -> number1.times(number2)
}
```

Kotlin high order function을 사용해 리팩토링 해보시면 좋을 것 같네요.

```kotlin
enum class Operator(val op: String, val fun: (a: Double, b: Double) -> Double) {
```

kyucumber commented on src/test/kotlin/calculator/MethodTest.kt:
```diff
@@ -0,0 +1,70 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class MethodTest {
```
이 테스트는 Operator와 관련된 테스트나 stdlib와 관련된 부분을 테스트하고 있네요. Operator와 관련된 테스트는 OperatorTest에 정의하고 다른 replace나 joinToString같은 stdlib는 테스트하지 않아도 괜찮을 것 같아요.

kyucumber commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,14 @@
+
+    companion object {
+        private val map = Operator.values().associateBy { it.operator }
+        fun find(operator: String) = map[operator]!!
```
연산자로 enum을 찾을 수 없는 경우에 NullPointerException이 발생하는게 적절할까요? 해당하는 연산자를 찾을 수 없다는 등의 메시지와 함께 다른 적절한 예외가 발생하도록 변경해주시면 어떨까요?

kyucumber reviewed (changes requested):
안녕하세요 몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

jysohn0825 committed:
Fix Operaotr enum

- fix operaotr enum method (find method occur IllegalArguementException, add high order function)
- change validCheck method, not if branch but require method

jysohn0825 committed:
Revert "Fix Operaotr enum"

This reverts commit 7d92bc5eac560224e6b10632dba70cb33d3645a1.

jysohn0825 committed:
Merge commit 'f8041e14195d5728497f237ded9f2430e9640b96' into step2

jysohn0825 committed:
Fix Operator enum

change operaotr enum (add high order function, add Exception)
change method name
add test code

kyucumber commented on src/test/kotlin/calculator/OperatorTest.kt:
```diff
@@ -0,0 +1,20 @@
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.shouldBe
+
+class OperatorTest : StringSpec({
```
kotest 👍

kyucumber reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 머지할게요~
